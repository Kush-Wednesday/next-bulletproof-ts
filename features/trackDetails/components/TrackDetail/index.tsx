import { If, T, CustomCard } from "../../../../common";
import { Button, Card as AntDCard, Tag, Image } from "antd";
import { isEmpty } from "lodash";
import { useRouter } from "next/router";
import React, { useRef, useState } from "react";
import { IntlShape, injectIntl } from "react-intl";
import styled from "styled-components";
import { TrackItem } from "../../api/getTrackDetails";
import { styles, colors, fonts } from "@themes";

type ContainerProps = {
  maxwidth: number;
  padding: number;
};

const Container = styled.div<ContainerProps>`
  && {
    display: flex;
    flex-direction: column;
    max-width: ${props => props.maxwidth}px;
    width: 100%;
    margin: 0 auto;
    padding: ${props => props.padding}px;
    background-color: "#57CC99";
  }
`;

const CustomTrackCard = styled(AntDCard)`
  && {
    padding: 1rem;
    border: 1px solid #111;
    border-radius: 10px;
    margin: 20px 0;
    background-color: ${colors.secondary};
    box-shadow: 0 0 5px rgba(2, 2, 2, 0.2);
  }
`;
const Flex = styled.div`
  display: flex;
  align-items: center;
  margin-bottom: 20px;
`;

const FColumn = styled.div`
  flex: 1;
  display: flex;
  justify-content: center;
  flex-direction: column;
`;
const StyledAudio = styled.div`
  
  margin: 0.5rem auto;
  background-color: ${props => (props.theme === "dark" ? "#222" : "#f5f5f5")};'

`;

interface TrackDetailProps {
  intl: IntlShape;
  trackDetails: TrackItem;
}
const TrackDetail: React.FC<TrackDetailProps> = ({ intl, trackDetails }) => {
  // console.log("trackdetails", trackDetails);
  const {
    artistName,
    collectionName,
    trackName,
    collectionPrice,
    trackPrice,
    artworkUrl100,
    previewUrl,
    trackId,
    currency,
    releaseDate,
  } = trackDetails;

  const [isPlaying, setIsPlaying] = useState(false);
  const router = useRouter();
  const audioRef = useRef<HTMLAudioElement>(null);

  function handleOnPlay(evt) {
    evt.preventDefault();
    if (isPlaying) {
      audioRef.current?.pause();
      setIsPlaying(false);
    } else {
      audioRef.current?.play();
      setIsPlaying(true);
    }
  }

  return (
    <Container data-testid="track-details" maxwidth={600} padding={20}>
      <Button type="primary" size="small" onClick={() => router.push("/")}>
        {intl.formatMessage({ id: "back_to_home_button" })}
      </Button>

      <CustomTrackCard>
        <Flex>
          <If
            condition={!isEmpty(artworkUrl100)}
            otherwise={<T data-testid="artwork-unavailable" id="artwork_unavailable" />}
          >
            <Image
              data-testid="album-art"
              style={{ paddingRight: "20px" }}
              height={200}
              src={artworkUrl100}
              alt="no image"
            />
          </If>

          <FColumn>
            <If
              condition={!isEmpty(collectionName)}
              otherwise={
                <T data-testid="collection-name-unavailable" id="collection_name_unavailable" />
              }
            >
              <T
                data-testid="collection-name"
                id="collection_name"
                values={{ collectionName: collectionName }}
              />
            </If>

            <StyledAudio>
              <audio
                id={previewUrl}
                data-testid="audio-test"
                aria-label="audio-label"
                autoPlay={false}
                controls
                onPlay={handleOnPlay}
                ref={audioRef}
                style={
                  {
                    display: "block",
                    width: "fit-content<100%>",
                    maxWidth: "100%",
                  } as React.CSSProperties
                }
              >
                <source src={previewUrl} type="audio/mp3"></source>
                Your browser does not support audio tags
              </audio>
            </StyledAudio>
          </FColumn>
        </Flex>

        <If
          condition={!isEmpty(artistName)}
          otherwise={<T data-testid="artist-name-unavailable" id="artist_name_unavailable" />}
        >
          <T data-testid="artist-name" id="artist_name" values={{ name: artistName }} />
        </If>

        <If condition={!isEmpty(trackName)}>
          <T text={trackName} type="heading" />
        </If>
        <If
          condition={!isNaN(collectionPrice)}
          otherwise={
            <T data-testid="collection-price-unavailable" id="collection_price_unavailable" />
          }
        >
          <T
            data-testid="collection-price"
            id="collection_price"
            values={{ currency: currency, collectionPrice: collectionPrice }}
          />
        </If>
        <If
          condition={!isEmpty(releaseDate)}
          otherwise={<T data-testid="release-date-unavailable" id="release_date_unavailable" />}
        >
          <T
            data-testid="release-date"
            id="release_date"
            values={{ releaseDate: new Date(releaseDate).toDateString() }}
          />
        </If>
      </CustomTrackCard>
    </Container>
  );
};

export default injectIntl(TrackDetail);
