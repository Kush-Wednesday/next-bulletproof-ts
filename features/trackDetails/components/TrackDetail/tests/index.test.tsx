import { fireEvent, render } from "@utils/testUtils";
import TrackDetail from "../index";

describe("<TrackCard />", () => {
  const trackDetailProps = {
    trackDetails: {
      artistName: "Adele",
      trackName: "Hello",
      trackId: 1544494392,
      artworkUrl100:
        "https://is5-ssl.mzstatic.com/image/thumb/Music116/v4/5a/47/6d/5a476ddd-4690-1297-1896-6a286a497a21/191404113974.png/100x100bb.jpg",
      previewUrl: "22233223",
      collectionName: "45",
      collectionPrice: 0.09,
      currency: "USD",
      releaseDate: "24 NOV 2018",
      trackPrice: 0.08,
    },
  };

  it("should match the snapshot", () => {
    const { baseElement } = render(<TrackDetail {...trackDetailProps} />);
    expect(baseElement).toMatchSnapshot();
  });

  it("should contain 1 RepoList component", () => {
    const { getAllByTestId } = render(<TrackDetail {...trackDetailProps} />);
    expect(getAllByTestId("track-details").length).toBe(1);
  });

  it("should check if image is rendered", () => {
    const { getByAltText } = render(<TrackDetail {...trackDetailProps} />);
    const a = getByAltText(/no image/i) as HTMLImageElement;
    expect(a.src).toContain(
      "https://is5-ssl.mzstatic.com/image/thumb/Music116/v4/5a/47/6d/5a476ddd-4690-1297-1896-6a286a497a21/191404113974.png/100x100bb.jpg"
    );
  });

  it("should render unavailable texts if props are undefined ", () => {
    const trackDetailProps1 = {
      trackDetails: {
        artistName: "Adele",
        trackName: "Hello",
        trackId: 1544494392,
        artworkUrl100:
          "https://is5-ssl.mzstatic.com/image/thumb/Music116/v4/5a/47/6d/5a476ddd-4690-1297-1896-6a286a497a21/191404113974.png/100x100bb.jpg",
        previewUrl: "22233223",
        collectionName: "",
        collectionPrice: 0.09,
        currency: "USD",
        releaseDate: "24 NOV 2018",
        trackPrice: 0.08,
      },
    };
    const { getAllByTestId } = render(<TrackDetail {...trackDetailProps1} />);

    expect(getAllByTestId("artist-name")[0].innerHTML).toContain("Adele");
    expect(getAllByTestId("collection-name-unavailable").length).toBe(1);
    //   expect(getAllByTestId('collection-name')).toBeFalsy;
  });
});
