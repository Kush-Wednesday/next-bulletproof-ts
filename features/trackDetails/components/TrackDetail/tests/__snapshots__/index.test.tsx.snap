// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TrackCard /> should match the snapshot 1`] = `
<body>
  <div>
    <div
      class="TrackDetail__Container-sc-cov9jc-0 booLkx"
      data-testid="track-details"
    >
      <button
        class="ant-btn ant-btn-primary ant-btn-sm"
        type="button"
      >
        <span>
          Back to Search
        </span>
      </button>
      <div
        class="ant-card ant-card-bordered TrackDetail__CustomTrackCard-sc-cov9jc-1 cIQhtJ"
      >
        <div
          class="ant-card-body"
        >
          <div
            class="TrackDetail__Flex-sc-cov9jc-2 curJWl"
          >
            <div
              class="ant-image"
              data-testid="album-art"
              style="height: 200px;"
            >
              <img
                alt="no image"
                class="ant-image-img"
                src="https://is5-ssl.mzstatic.com/image/thumb/Music116/v4/5a/47/6d/5a476ddd-4690-1297-1896-6a286a497a21/191404113974.png/100x100bb.jpg"
                style="height: 200px; padding-right: 20px;"
              />
              <div
                class="ant-image-mask"
              >
                <div
                  class="ant-image-mask-info"
                >
                  <span
                    aria-label="eye"
                    class="anticon anticon-eye"
                    role="img"
                  >
                    <svg
                      aria-hidden="true"
                      data-icon="eye"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M942.2 486.2C847.4 286.5 704.1 186 512 186c-192.2 0-335.4 100.5-430.2 300.3a60.3 60.3 0 000 51.5C176.6 737.5 319.9 838 512 838c192.2 0 335.4-100.5 430.2-300.3 7.7-16.2 7.7-35 0-51.5zM512 766c-161.3 0-279.4-81.8-362.7-254C232.6 339.8 350.7 258 512 258c161.3 0 279.4 81.8 362.7 254C791.5 684.2 673.4 766 512 766zm-4-430c-97.2 0-176 78.8-176 176s78.8 176 176 176 176-78.8 176-176-78.8-176-176-176zm0 288c-61.9 0-112-50.1-112-112s50.1-112 112-112 112 50.1 112 112-50.1 112-112 112z"
                      />
                    </svg>
                  </span>
                  Preview
                </div>
              </div>
            </div>
            <div
              class="TrackDetail__FColumn-sc-cov9jc-3 gYfbkr"
            >
              <p
                class="T__StyledText-sc-apdcd0-0 fjJkvM"
                data-testid="collection-name"
              >
                Collection Name : 45 
              </p>
              <div
                class="TrackDetail__StyledAudio-sc-cov9jc-4 dOyzXU"
              >
                <audio
                  aria-label="audio-label"
                  controls=""
                  data-testid="audio-test"
                  id="22233223"
                  style="display: block; max-width: 100%;"
                >
                  <source
                    src="22233223"
                    type="audio/mp3"
                  />
                  Your browser does not support audio tags
                </audio>
              </div>
            </div>
          </div>
          <p
            class="T__StyledText-sc-apdcd0-0 fjJkvM"
            data-testid="artist-name"
          >
            Artist Name: Adele
          </p>
          <p
            class="T__StyledText-sc-apdcd0-0 bPJaah"
            data-testid="t"
          >
            Hello
          </p>
          <p
            class="T__StyledText-sc-apdcd0-0 fjJkvM"
            data-testid="collection-price"
          >
            Collection Price : 0.09 USD
          </p>
          <p
            class="T__StyledText-sc-apdcd0-0 fjJkvM"
            data-testid="release-date"
          >
            Release Date: Sat Nov 24 2018
          </p>
        </div>
      </div>
    </div>
  </div>
</body>
`;
